lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,688 @@
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitMetatypeType(MetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitExistentialMetatypeType(ExistentialMetatypeType *type) {
+  return getClangMetatypeType(ClangASTContext);
+}
+
+clang::QualType ClangTypeConverter::visitClassType(ClassType *type) {
+  auto &clangCtx = ClangASTContext;
+  // produce the clang type INTF * if it is imported ObjC object.
+  auto swiftDecl = type->getDecl();
+  if (swiftDecl->isObjC()) {
+    clang::IdentifierInfo *ForwardClassId =
+      &clangCtx.Idents.get(swiftDecl->getName().get());
+    auto *CDecl = clang::ObjCInterfaceDecl::Create(
+                          clangCtx, clangCtx.getTranslationUnitDecl(),
+                          clang::SourceLocation(), ForwardClassId,
+                          /*typeParamList*/nullptr, /*PrevDecl=*/nullptr,
+                          clang::SourceLocation());
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this class.
+    SmallString<64> runtimeNameBuffer;
+    CDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     CDecl->getASTContext(),
+                     swiftDecl->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCInterfaceType(CDecl);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+  return getClangIdType(clangCtx);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
+
+  if (type->getDecl()->getName().str() == "Optional") {
+    auto args = type->getGenericArgs();
+    if (args.size() != 1)
+      // We have a weird Optional with multiple generic arguments, eh?
+      return clang::QualType();
+    return convert(args[0]);
```

Comment:
@rjmccall what should we be doing here? This seems very wrong, but I wasn't sure what to do. We cannot simply reuse the code from `GenClangType`.

```
  if (auto underlyingTy =
          SILType::getPrimitiveObjectType(type).getOptionalObjectType()) {
    // The underlying type could be a bridged type, which makes any
    // sort of casual assertion here difficult.
    return Converter.convert(IGM, underlyingTy.getASTType());
  }
```

Reply:
The only Optional types that have Clang equivalents are those with single-class-reference representations; there's a method around somewhere for that.

Reply:
> The only Optional types that have Clang equivalents are those with single-class-reference representations; there's a method around somewhere for that.

I found `ImportName.cpp/canImportAsOptional` which is probably the one 🙂. There's also `DeclAndTypePrinter::isClangPointerType` which has a similar check.